### PR TITLE
feature: ProjectWorker

### LINKED ISSUE #36: add support for Project Worker
Add support to run project submissions in frameworks like react, node, django etc.

Refer [https://github.com/coding-blocks/judge-taskmaster/pull/35](this) pull request for the initial work that has been done. Create your PR against the `project-worker` branch

- [ ] Create project scenario
    - [ ] Complete setup method
    - [ ] Complete run method
    - [ ] Complete result method
    - [ ] Write tests for the methods and scenario
- [ ]  Refactor `src/tasks/job.ts`
    - [ ] Create `src/tasks/jobs/` directory
    - [ ] Move all the jobs other than Base class to individual files in jobs dir 
           Like move `RunJob` and its type to `src/tasks/jobs/run.ts`
    - [ ] Refactor imports to support current structure

### PR DESCRIPTION
**Pull request by prabalsingh24**
fixes #36, #39 

- [x] Create project scenario

  - [x]   Complete setup method ( as done here in the test)
  - [x]   Complete run method
    - [x] Complete result method
    - [ ] Write tests for the methods and scenario

- [x] Send the payload to the job

- [x] Refactor src/tasks/job.ts

    - [x] Create src/tasks/jobs/ directory
   - [x] Move all the jobs other than Base class to individual files in jobs dir
   - [x]  Like move RunJob and its type to src/tasks/jobs/run.ts
    - [x] Refactor imports to support current structure

### CODE REVIEW

**Comment by prabalsingh24** [comment]
@jatinkatyal13 can you have a look and let me know if I am not the correct path.

Also how will the score be calculated?

**Comment by prabalsingh24** [comment]
Also how to extract zip file. I assume a library will be used for that. Let me know if you have any particular lib in your mind

**Comment by prabalsingh24** [comment]
This seems to be workingg. 

Screenshot from 2020-05-23 19-33-49

**Comment by prabalsingh24** [comment]
Can you tell me what will be returned in which cases. (build fail, complie fail, etc ) what scores and code should be returned?

**Comment by prabalsingh24** [comment]
Do I need to check how many tests are passing and how many are failing?

**Comment by prabalsingh24** [comment]
Screenshot from 2020-05-23 21-39-57

**Comment by prabalsingh24** [comment]
TODOs: 
1. Instead of using zip file locally. Download it and then extract it. 
2. Modify project result function to return correct `code` and `score`
3. Add `LANG.CONFIG` 
4. Write tests :P

File: src/tasks/scenarios/project.ts [hunk 1]
```diff
@@ -1,32 +1,80 @@
...
+
+    const solutionDir = path.join(currentJobDir, 'solution')
+    mkdir('-p', solutionDir)
+    solutionZip.extractAllTo(solutionDir, true);
+  }
+
+  run(currentJobDir: string, job: ProjectJob) {
+
+    // LANG_CONFIG is undefined rn, hence card coding the value of cpus and memory
+    const LANG_CONFIG = config.LANGS[job.lang]
```
Comment (jatinkatyal13) [inline src/tasks/scenarios/project.ts:29]: We will be creating a new config here named PROJECTS. In future, we might need more configurable information for these projects.

File: src/tasks/scenarios/project.ts [hunk 2]
```diff
@@ -1,32 +1,80 @@
...
 export default class ProjectScenario extends Scenario {
   async setup(currentJobDir: string, job: ProjectJob) {
-    // TODO
-    const problemDir = path.join(currentJobDir, job.problem)
-    await download(job.problem, problemDir)
 
-    const solutionDir = path.join(currentJobDir, job.source)
-    await download(job.problem, solutionDir)
+    // for testing only, change it to download
+    var problemZip = new AdmZip('/judge-worker-prabal/current/problem.zip')
```
Comment (jatinkatyal13) [inline src/tasks/scenarios/project.ts:14]: We were already downloading the zip. Using `exec('unzip abc.zip')` should work

File: src/tasks/scenarios/project.ts [hunk 3]
```diff
@@ -1,32 +1,80 @@
...
+      }
+    }
+
+    const build_stderr = cat(path.join(currentJobDir, 'build.stderr')).toString()
+    if (build_stderr) {
+      return {
+        id: job.id,
+        stderr: build_stderr,
+        stdout: '',
+        code: 12123,
```
Comment (jatinkatyal13) [inline src/tasks/scenarios/project.ts:61]: set code to `result_code`, time to `result.time` and score to 0

File: src/tasks/scenarios/project.ts [hunk 4]
```diff
@@ -1,32 +1,80 @@
...
+    const result_code = cat(path.join(currentJobDir, 'result.code')).toString()
+    if (result_code) {
+      // problem hash and solution hash were not equal. // error
+      return {
+        id: job.id,
+        stderr: cat(path.join(currentJobDir, 'result.stderr')).toString(),
+        stdout: '',
+        code: parseInt(result_code),
+        time: 1,
+        score: 12
```
Comment (jatinkatyal13) [inline src/tasks/scenarios/project.ts:51]: score will be 0

**Comment by jatinkatyal13** [comment]
Nice job!

So here are few pointers
1. The score will 100 if and only if `result.code` file if 0
2. For unzipping the downloaded zip we can use the Linux unzip command using the `exec` from shelljs
3. The score is currently binary, you either get 100 or 0
4. Yes the status code is right and generated from worker's runguard

**Comment by jatinkatyal13** [comment]
@prabalsingh24 are facing any issue with the changes requested? You can always reach out to me on discord for any help. Let's complete this task and we'll be ready to deploy the new judge.

**Comment by prabalsingh24** [comment]
> @prabalsingh24 are facing any issue with the changes requested? You can always reach out to me on discord for any help. Let's complete this task and we'll be ready to deploy the new judge.

Yeah thankss. Actually I didn't work yesterday :P I am working on it rn. I will push changes soon :)

File: src/tasks/scenarios/project.ts [hunk 5]
```diff
@@ -1,32 +1,77 @@
...
+        id: job.id,
+        stderr: cat(path.join(currentJobDir, 'result.stderr')).toString(),
+        stdout: '',
+        code: parseInt(result_code),
+        time: 0,
+        score: 0
+      }
+    }
+
+    // if hashes were equal, result_code will be found inside solution directory
```
Comment (prabalsingh24) [inline src/tasks/scenarios/project.ts:51]: @jatinkatyal13 is this done on purpose? `result_code` and `result_time` inside solutionDirectory? and other stuff in the baseDirectory?

**Comment by prabalsingh24** [comment]
Screenshot from 2020-05-25 17-38-32

File: src/tasks/scenarios/project.ts (hunk 5 above)
Comment (jatinkatyal13) [reply src/tasks/scenarios/project.ts:51]: Nice catch, I'll fix this [here](https://github.com/coding-blocks/judge-workers/blob/project-worker/containers/project-worker/base/judge.sh#L32). You can safely assume everything related to the result will be in the base directory

**Pull request merged**